File: web/src/session/turnMentions.test.ts
import { expect, it } from 'vitest'
import { extractExpertRefNames, parseTurnMentions } from './turnMentions'
import vectors from './turnMentions.vectors.json'

const ID = '01ARZ3NDEKTSV4RRFFQ69G5FAV'

it('parses expert tokens and @ names as the same actor', () => {
  const token = parseTurnMentions(`[引用专家 安全工程师|${ID}] 请审一下`)
  const at = parseTurnMentions('@安全工程师 请审一下')
  expect(token).toEqual([{ kind: 'expert', id: ID, name: '安全工程师' }])
  expect(at).toEqual([{ kind: 'member', id: '', name: '安全工程师' }])
  expect(extractExpertRefNames(`[引用专家 安全工程师|${ID}] 请审一下`)).toEqual(['安全工程师'])
})

it('matches the shared C4 mention vectors', () => {
  for (const vec of vectors) {
    expect(parseTurnMentions(vec.text), vec.name).toEqual(vec.want)
  }
})

File: internal/app/turn_mentions_test.go
package app

import "testing"

func TestParseTurnMentionsSameActor(t *testing.T) {
	id := "01ARZ3NDEKTSV4RRFFQ69G5FAV"
	token := ParseTurnMentions("[引用专家 安全工程师|" + id + "] 请审一下")
	at := ParseTurnMentions("@安全工程师 请审一下")
	if len(token) != 1 || token[0].Kind != "expert" || token[0].ID != id || token[0].Name != "安全工程师" {
		t.Fatalf("token = %#v", token)
	}
	if len(at) != 1 || at[0].Kind != "member" || at[0].Name != "安全工程师" {
		t.Fatalf("at = %#v", at)
	}
	if extractExpertRefNames("[引用专家 安全工程师|" + id + "] 请审一下")[0] != "安全工程师" {
		t.Fatal("extractExpertRefNames must keep the display name")
	}
}
